applications/ParticleMechanicsApplication/custom_conditions/particle_based_conditions/mpm_particle_penalty_coupling_interface_condition.cpp

```diff
@@ -0,0 +1,238 @@
+        const unsigned int number_of_nodes = rGeom.PointsNumber();
+
+        // At the beginning of NonLinearIteration, REACTION has to be reset to zero
+        for ( unsigned int i = 0; i < number_of_nodes; i++ )
+        {
+            rGeom[i].SetLock();
+            rGeom[i].FastGetSolutionStepValue(REACTION).clear();
+            rGeom[i].UnSetLock();
+        }
+
+        mReactionIsAdded = false;
+    }
+}
+
+//************************************************************************************
+//************************************************************************************
+
+void MPMParticlePenaltyCouplingInterfaceCondition::FinalizeNonLinearIteration(ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY
+
+    // Estimating the contact forces at the boundary
+    if (Is(INTERFACE))
+    {
+        this->CalculateInterfaceContactForce( rCurrentProcessInfo );
+    }
+
+    KRATOS_CATCH( "" )
+}
+
+
+void MPMParticlePenaltyCouplingInterfaceCondition::FinalizeSolutionStep( ProcessInfo& rCurrentProcessInfo )
+{
+    KRATOS_TRY
+
+    MPMParticlePenaltyDirichletCondition::FinalizeSolutionStep(rCurrentProcessInfo);
+
+    // Estimating the contact forces at the boundary
+    if (Is(INTERFACE))
+    {
+        this->CalculateInterfaceContactForce( rCurrentProcessInfo );
+    }
+
+    KRATOS_CATCH( "" )
+}
+
+//************************************************************************************
+//************************************************************************************
+
+void MPMParticlePenaltyCouplingInterfaceCondition::CalculateAll(
+    MatrixType& rLeftHandSideMatrix, VectorType& rRightHandSideVector,
+    ProcessInfo& rCurrentProcessInfo,
+    bool CalculateStiffnessMatrixFlag,
+    bool CalculateResidualVectorFlag
+    )
+{
+    KRATOS_TRY
+
+    MPMParticlePenaltyDirichletCondition::CalculateAll(rLeftHandSideMatrix, rRightHandSideVector,
+                                                        rCurrentProcessInfo, CalculateStiffnessMatrixFlag,
+                                                        CalculateResidualVectorFlag);
+
+    // Append penalty force at the nodes
+    if (Is(INTERFACE) && !mReactionIsAdded)
+    {
+        this->CalculateNodalContactForce( rRightHandSideVector, rCurrentProcessInfo, CalculateResidualVectorFlag );
+        mReactionIsAdded = true;
+    }
+
+    KRATOS_CATCH( "" )
+}
+
+//************************************************************************************
+//************************************************************************************
+
+void MPMParticlePenaltyCouplingInterfaceCondition::CalculateNodalContactForce( const VectorType& rRightHandSideVector, ProcessInfo& rCurrentProcessInfo, const bool CalculateResidualVectorFlag )
+{
+    if ( CalculateResidualVectorFlag == true )
+    {
+        GeometryType& rGeom = GetGeometry();
+        const unsigned int number_of_nodes = rGeom.size();
+        const unsigned int dimension = rGeom.WorkingSpaceDimension();
```

Comment:
dimension is actually another variable concerning the geometry dimension i.e. curve is 1 surface is 2, ....

Reply:
Yes, but what I mean here is the space dimension, not the geometry dimension; referring to `Dimension()`, `WorkingSpaceDimension()`, and `LocalSpaceDimension()` in `geometry_data.h`

Reply:
Yes, I know. You're using it correctly. Only the name is confusing.